src/renderer/components/Editor/TakeGroupComponent.tsx

```diff
@@ -91,7 +90,38 @@ const TakeGroupComponent = ({
   };
 
   const ungroupTakes = () => {
-    dispatch(deleteTakeGroup(takeGroup.id));
+    const getTakeRanges = (): IndexRange[] => {
```

Comment:
This should definitely be a util function instead of being inline in the react component.

I tried to refactor it a little to make it easier to read, because I was struggling to understand the logic. In order to minimise usage of mutating variables (which tends to be the biggest reason functions get hard to follow), I ended up rewriting it entirely as a `reduce`:

```
const getTakeRanges: () => IndexRange[] = () => {
  const { words } = transcription;

  // Make a list of only the words in this take group,
  // keeping track of what their indices in the current transcription were
  const wordsInTakeGroup: {
    word: Word;
    indexInTranscription: number;
  }[] = words
    // Keep track of the indices in the transcription so that we can use them later
    .map((word, i) => ({ word, indexInTranscription: i }))
    // Filter to only include words in the current take group
    .filter(({ word }) => word.takeInfo?.takeGroupId === takeGroup.id);

  // Extract a list of take ranges from the word list we just made
  const takeRanges = wordsInTakeGroup.reduce<IndexRange[]>(
    (rangesSoFar, word, index) => {
      const isLastWordInTakeGroup = index === wordsInTakeGroup.length - 1;
      const nextWord = isLastWordInTakeGroup ? null : wordsInTakeGroup[index + 1];
      const isLastWordInTake =
        nextWord?.word.takeInfo?.takeIndex !==
        word.word.takeInfo?.takeIndex;

      // If we're in a new take compared to the previous word, or this is the last
      // word in the take group, 'push' the current take to the list of takes.
      if (isLastWordInTakeGroup || isLastWordInTake) {
        // Find what the startIndex of the take should be. Since takes are always
        // next to each other, the start index will just be the endIndex of the previous
        // take range, or if this is the first take in the group, it will be the index of
        // the first word in the take.
        const startIndex =
          rangesSoFar.length > 0
            ? rangesSoFar[rangesSoFar.length - 1].endIndex
            : wordsInTakeGroup[0].indexInTranscription;

        // Add a new range from startIndex up to the current word's index in the current
        // transcription, plus one as the end index is exclusive
        return rangesSoFar.concat([
          {
            startIndex,
            endIndex: word.indexInTranscription + 1,
          },
        ]);
      }

      // If this isn't the last word or a new take, just return the list of ranges unmodified
      return rangesSoFar;
    },
    []
  );

  return takeRanges;
};
```

I think this is a lot easier to reason about, even if it looks more verbose from the comments. Feel free to use this or to take the ideas and include them in your code. Also note that regardless of what the final function ends up being, this is exactly the kind of thing that needs unit tests so please write some. I've tested this with Max's sauce video but haven't tested three-take groups or anything more complex.

Reply:
Yup that makes sense. Thanks for translating it into a more functional programming style, I wasn't sure how necessary it was so I just caved into imperative rip.

I'll make some test cases up next, definitely a good idea.